### PR TITLE
refactor task engine unit tests

### PR DESCRIPTION
**Pull request by aaithal**
### Summary

refactor task engine unit tests

### Implementation details

This commit moves commonly used code in the task engine unit tests
into methods. It also reuses these methods in various places to
increase the readability of the tests. It also deletes a number of
redundant tests and adds a new table based test for the happy path
test.

Test covered has now increased to 79.3% from 77.8% for this package

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

File: agent/engine/common_test.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
+// Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (richardpen) [inline agent/engine/common_test.go]: This should be updated to 2018?

File: agent/engine/common_test.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	}()
+	return func() {
+		done <- true
+	}
+}
+
+func verifyTaskIsRunning(stateChangeEvents <-chan statechange.Event, testTasks ...*api.Task) error {
+	for {
+		select {
+		case event := <-stateChangeEvents:
```
Comment (richardpen) [inline agent/engine/common_test.go]: Can we get rid of the `select` statement, seems we have only one case here? Also below in `verifyTaskIsStopped`

File: agent/engine/common_test.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+	assertions func(),
+) {
+	imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes()
+	client.EXPECT().PullImage(container.Image, nil).Return(DockerContainerMetadata{})
+	imageManager.EXPECT().RecordContainerReference(container).Return(nil)
+	imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).Return(nil)
+	client.EXPECT().APIVersion().Return(defaultDockerClientAPIVersion, nil)
+	dockerConfig, err := task.DockerConfig(container, defaultDockerClientAPIVersion)
+	if err != nil {
+		t.Fatal(err)
```
Comment (richardpen) [inline agent/engine/common_test.go]: Can you change this to use `require`?

File: agent/engine/common_test.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	assert.Equal(t, event.(api.ContainerStateChange).Status, api.ContainerRunning,
+		"Expected container to be RUNNING")
+
+	event = <-stateChangeEvents
+	assert.Equal(t, event.(api.TaskStateChange).Status, api.TaskRunning,
+		"Expected task to be RUNNING")
+
+	select {
+	case <-stateChangeEvents:
+		t.Fatal("Should be out of events")
```
Comment (richardpen) [inline agent/engine/common_test.go]: Can you change this to use `require`?

File: agent/engine/common_test.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+			assert.Equal(t, *cont.ExitCode, 1, "Exit code should be present")
+		}
+	}
+
+	event = <-stateChangeEvents
+	assert.Equal(t, event.(api.TaskStateChange).Status, api.TaskStopped, "Expected task to be STOPPED")
+
+	select {
+	case <-stateChangeEvents:
+		t.Fatal("Should be out of events")
```
Comment (richardpen) [inline agent/engine/common_test.go]: Same as above.

File: agent/engine/docker_task_engine_test.go [hunk 6]
```diff
@@ -810,104 +401,50 @@ func TestRemoveEvents(t *testing.T) {
...
-			func(config *docker.Config, y interface{}, containerName string, z time.Duration) {
-				createdContainerName = containerName
-				createStartEventsReported.Add(1)
-				go func() {
-					eventStream <- createDockerEvent(api.ContainerCreated)
-					createStartEventsReported.Done()
-				}()
-			}).Return(DockerContainerMetadata{DockerID: containerID})
+	containerName := make(chan string)
+	go func() {
```
Comment (richardpen) [inline agent/engine/docker_task_engine_test.go]: Can this be just `go setCreatedContainerName(<-containerName)`?

File: agent/engine/docker_task_engine_test.go [hunk 7]
```diff
@@ -99,513 +119,123 @@ func mockSetupNSResult() *current.Result {
...
+
+			eventStream := make(chan DockerContainerChangeEvent)
+			// createStartEventsReported is used to force the test to wait until the container created and started
+			// events are processed
+			createStartEventsReported := sync.WaitGroup{}
+
+			client.EXPECT().Version().Return("1.12.6", nil)
+			client.EXPECT().ContainerEvents(gomock.Any()).Return(eventStream, nil)
+			containerName := make(chan string)
+			go func() {
```
Comment (richardpen) [inline agent/engine/docker_task_engine_test.go]: Can this be just `go setCreatedContainerName(<-containerName)`?

File: agent/engine/common_test.go (hunk 1 above)
Comment (aaithal) [reply agent/engine/common_test.go]: yes, sure.

File: agent/engine/common_test.go (hunk 2 above)
Comment (aaithal) [reply agent/engine/common_test.go]: I'm not sure why it was written this way. I'll try, but revert to this if it starts breaking things

File: agent/engine/common_test.go (hunk 3 above)
Comment (aaithal) [reply agent/engine/common_test.go]: We're explicitly not doing `require` here. There's some weirdness in the return type, which doesn't treat `nil` value for this type as `nil`.

File: agent/engine/docker_task_engine_test.go (hunk 7 above)
Comment (aaithal) [reply agent/engine/docker_task_engine_test.go]: yeah, sure

File: agent/engine/docker_task_engine_test.go [hunk 8]
```diff
@@ -52,18 +50,46 @@ const (
...
 var (
-	defaultConfig                 = config.DefaultConfig()
-	nsResult                      = mockSetupNSResult()
-	defaultDockerClientAPIVersion = dockerclient.Version_1_17
+	defaultConfig config.Config
+	nsResult      = mockSetupNSResult()
+
+	// createdContainerName is used to save the name of the created
+	// container from the validateContainerRunWorkflow method. This
+	// variable should never be access directly.
```
Comment (sharanyad) [inline agent/engine/docker_task_engine_test.go]: nit: accessed

File: agent/engine/docker_task_engine_test.go [hunk 9]
```diff
@@ -99,513 +119,124 @@ func mockSetupNSResult() *current.Result {
...
-		// Container config should get updated with this during CreateContainer
-		dockerConfig.Labels["com.amazonaws.ecs.task-arn"] = sleepTask.Arn
-		dockerConfig.Labels["com.amazonaws.ecs.container-name"] = container.Name
-		dockerConfig.Labels["com.amazonaws.ecs.task-definition-family"] = sleepTask.Family
-		dockerConfig.Labels["com.amazonaws.ecs.task-definition-version"] = sleepTask.Version
-		dockerConfig.Labels["com.amazonaws.ecs.cluster"] = ""
-		metadataManager.EXPECT().Create(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("create metadata error"))
-		client.EXPECT().CreateContainer(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Do(
-			func(config *docker.Config, y interface{}, containerName string, z time.Duration) {
+			// And then docker reports that sleep died, as sleep is wont to do
```
Comment (sharanyad) [inline agent/engine/docker_task_engine_test.go]: edit needed for this comment?

File: agent/engine/common_test.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+// create operation. It can be used to validate that the name of the container
+// removed matches with the generated container name during cleanup operation in the
+// test.
+func validateContainerRunWorkflow(t *testing.T,
+	container *api.Container,
+	task *api.Task,
+	imageManager *MockImageManager,
+	client *MockDockerClient,
+	roleCredentials *credentials.TaskIAMRoleCredentials,
+	createStartEventsReported sync.WaitGroup,
```
Comment (adnxn) [inline agent/engine/common_test.go]: nit: should we add `*WG`or `*WaitGroup` to these names for clarity?

**Comment by aaithal** [comment]
> Also I just realised we only use those `verifyTaskIs*` functions in our `*_integ` tests. I wonder if we can verify those things some other way without having to dig into `stateChangeEvents` channel to validate.  

I completely missed that @adnxn. That seems like a worthwhile thing to do. May be as a follow up PR. I'll add a `TODO` for that.

File: agent/engine/common_test.go (hunk 10 above)
Comment (aaithal) [reply agent/engine/common_test.go]: I like the `WG` suffix idea. I'll do that for this PR and we can start enforcing that in the future, at least in tests where we use it heavily. Might lead to more readable code.

cc @sharanyad  @richardpen @yunhee-l

**Pull request merged**